File: web/src/features/credentialMeta.ts
/** Shared credential meta parsing (name / group / upstream_token_id / platform_user_id). */
export function parseCredentialMeta(metaJSON?: string): {
  name?: string;
  group?: string;
  upstream_token_id?: number;
  platform_user_id?: number;
  /** Raw parsed object, for callers that must round-trip unknown keys. */
  raw?: Record<string, unknown>;
} {
  if (!metaJSON?.trim()) return {};
  try {
    const parsed = JSON.parse(metaJSON) as Record<string, unknown>;
    const name = typeof parsed.name === "string" ? parsed.name : undefined;
    const group =
      typeof parsed.group === "string"
        ? parsed.group
        : typeof parsed.Group === "string"
          ? parsed.Group
          : undefined;
    const upstream =
      typeof parsed.upstream_token_id === "number"
        ? parsed.upstream_token_id
        : undefined;
    return {
      name,
      group,
      upstream_token_id: upstream,
      platform_user_id: positiveInt(parsed.platform_user_id),
      raw: parsed,
    };
  } catch {
    return {};
  }
}

/** positiveInt coerces a number or a numeric string ("1544") to a positive int. */
function positiveInt(value: unknown): number | undefined {
  const parsed =
    typeof value === "number"
      ? value
      : typeof value === "string" && value.trim() !== ""
        ? Number(value.trim())
        : Number.NaN;
  return Number.isInteger(parsed) && parsed > 0 ? parsed : undefined;
}

/**
 * withCredentialMetaValue merges one key into an existing credential meta
 * document, preserving every other key verbatim.
 *
 * Store-owned keys (`checkin_path` / `checkin_method` / `headers` for external
 * check-in, `name` / `group` / `upstream_token_id` for API keys) must survive a
 * frontend round-trip, so the whole object is carried rather than rebuilt.
 * `undefined` removes the key; removing the last key yields `"{}"` (an empty
 * object) rather than `""`, because the admin API treats an empty meta_json as
 * "field not supplied" and would otherwise ignore the removal.
 *
 * An unparsable existing document is discarded in favour of the single key:
 * writing a user id must not be blocked by legacy garbage.
 */
export function withCredentialMetaValue(
  metaJSON: string | undefined,
  key: string,
  value: number | string | undefined,
): string {
  const existing = parseCredentialMeta(metaJSON).raw ?? {};
  const next: Record<string, unknown> = { ...existing };
  if (value === undefined) delete next[key];
  else next[key] = value;
  return JSON.stringify(next);
}

File: internal/httpapi/credential_meta.go
package httpapi

import (
	"encoding/json"
	"errors"
	"io"
	"strconv"
	"strings"

	"github.com/lan/meta-gateway/internal/adapters"
)

// normalizeCredentialMeta validates and canonicalizes a credential meta_json
// document sent by the admin API.
//
// Only keys the gateway interprets are normalized; every other key is carried
// through verbatim so store-owned fields (external check-in checkin_path /
// checkin_method / headers, api-key name / group / upstream_token_id) survive a
// frontend round-trip. `platform_user_id` is rewritten as a bare JSON number
// because the New-API family user-id headers are derived from it, and imports
// may carry it as a quoted string.
//
// Empty input (or a document whose only key was an explicit null
// platform_user_id) returns "" — "no metadata".
func normalizeCredentialMeta(raw string) (string, error) {
	if strings.TrimSpace(raw) == "" {
		return "", nil
	}
	decoder := json.NewDecoder(strings.NewReader(raw))
	decoder.UseNumber()
	var fields map[string]json.RawMessage
	if err := decoder.Decode(&fields); err != nil || fields == nil {
		return "", errors.New("meta_json must be a JSON object")
	}
	// Reject trailing garbage after the object.
	var extra any
	if err := decoder.Decode(&extra); err != io.EOF {
		return "", errors.New("meta_json must be a JSON object")
	}
	if rawUserID, ok := fields["platform_user_id"]; ok {
		var value any
		if err := json.Unmarshal(rawUserID, &value); err != nil {
			return "", errors.New("platform_user_id must be a positive integer")
		}
		if value == nil {
			delete(fields, "platform_user_id")
		} else {
			id, ok := adapters.CoercePlatformUserID(value)
			if !ok {
				return "", errors.New("platform_user_id must be a positive integer")
			}
			fields["platform_user_id"] = json.RawMessage(strconv.FormatInt(id, 10))
		}
	}
	if len(fields) == 0 {
		return "", nil
	}
	encoded, err := json.Marshal(fields)
	if err != nil {
		return "", errors.New("meta_json must be a JSON object")
	}
	return string(encoded), nil
}
